WordPress/gutenberg
# Query block enhanced pagination: Detect inner plugin blocks during render

## What?



Fixes https://github.com/WordPress/gutenberg/issues/55706.

## Why?



The main problem is explained in the following issue:

- https://github.com/WordPress/gutenberg/issues/55706

## How?



By applying two changes:

- If there is a Pattern or Template Part inside the Query block, switch to "Auto" and inform the user.
- During the rendering, check that all the inner blocks are compatible. If we find one that it's not, disable the enhanced pagination.

I tried to use `ToggleGroupControl` instead of `ToggleControl` because it makes the change to "Auto" much more explicit, but `ToggleGroupControl` has a bug when the value is changed programmatically using `setAttributes()`. Also, `ToggleGroupControl` doesn't have support yet for _disabled_, which is required for this use case.

## Testing Instructions






- Add a Query block with some regular core blocks (group, featured image, post title…).
- Make sure that the enhanced pagination can be enabled as previously (no change).
- Enable it.
- Now add a synced Pattern or Template Part somewhere inside the Query block.
- You should see a modal informing that the enhanced pagination has switch to "Auto".
- Check the enhanced pagination setting, you should see a similar message in the Inspector Panel.
- Add another Pattern or Template Part. The modal should not appear again.
- Add a Post Content block.
- A new modal should appear informing you that the enhanced pagination has been disabled.
- Check the enhanced pagination setting, you should see a new message in the Inspector Panel.
- Now the enhanced pagination should not mention "Auto" and instead, be turned off and disabled.
- Remove the Post Content block.
- Check that the enhanced pagination has switched back to "Auto" and that it can be enabled again.
- Now repeat the last process using a block from an external plugin.
- You should see messages similar to the Post Content block.
- Check that adding more blocks don't trigger new modals, unless they are of different types and the settings changes from less restrictive to more restrive.

## Screenshots or screencast

I made a video to explain how the UX works, especially to get feedback from design about the "Auto" mode:

<div>
    <a href="https://www.loom.com/share/311ee1534a1641a786b765b6cbff48bb">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/311ee1534a1641a786b765b6cbff48bb-with-play.gif">
    </a>
  </div>

https://www.loom.com/share/311ee1534a1641a786b765b6cbff48bb

_[Drawings](https://excalidraw.com/#json=LujV45t88k1zcz2ZtIUku,nJNvbjk1uvKz4QlR-fcK_w)_

## Tasks

- [ ] Add PHP unit tests
- [ ] Get feedback on the texts
- [ ] Get feedback on the UX
- [ ] Make sure that we don't need to support `core/pattern`
